### PR TITLE
Partial FreeBSD Support

### PR DESCRIPTION
**Pull request by landonf**
This provides a non-controversial baseline for FreeBSD support.

Additional work will be required for the full port, but with this initial work -- plus local work-arounds for linker script behavior and use of pkg-installed libc++ -- I'm able to successfully compete a full Swift build on FreeBSD 10.x.

### CODE REVIEW

**Comment by carlossless** [comment]
:+1:

**Comment by jckarter** [comment]
Nice! @gribozavr is probably the best person to review the build system changes.

**Comment by gribozavr** [comment]
@landonf Great work!  Please update the pull request per comments above.  Which systems have you tested this branch on?

**Comment by landonf** [comment]
Thanks for the feedback; I believe I've integrated all the requested changes. I also moved ahead with a common Unix toolchain class pending additional comments on a preferred alternative.

I've only tested locally against x86-64-freebsd10.2 (-STABLE)

**Comment by gribozavr** [comment]
I'll run the tests on OS X and Linux, while waiting for feedback from @jrose-apple .

**Comment by gribozavr** [comment]
@landonf The new test fails on Linux.

```
FAIL: Swift :: BuildConfigurations/x64FreeBSDTarget.swift (141 of 7260)
******************** TEST 'Swift :: BuildConfigurations/x64FreeBSDTarget.swift' FAILED ********************
Script:
--
'/home/gribozavr/src.public/build/Ninja-ReleaseAssert/swift-linux-x86_64/bin/swift' -frontend -module-cache-path '/tmp/swift-testsuite-clang-module-cacheQ2TPZE' -disable-objc-attr-requires-foundation-module -parse /home/gribozavr/src.public/swift/test/BuildConfigurations/x64FreeBSDTarget.swift -verify -D FOO -D BAR -target x86_64-freebsd10 -disable-objc-interop -D FOO -parse-stdlib
'/home/gribozavr/src.public/build/Ninja-ReleaseAssert/swift-linux-x86_64/bin/swift-ide-test' -module-cache-path '/tmp/swift-testsuite-clang-module-cacheQ2TPZE' -completion-cache-path '/tmp/swift-testsuite-completion-cachee5rxPO' -test-input-complete -source-filename=/home/gribozavr/src.public/swift/test/BuildConfigurations/x64FreeBSDTarget.swift -target x86_64-freebsd10
--
Exit Code: 1

Command Output (stderr):
--
<unknown>:0: error: unsupported target OS: ''

--

********************
```

I think this is because the FreeBSD toolchain is being disabled on Linux with a `#if` in the C++ source.  Could you take a look?

**Comment by landonf** [comment]
@gribozavr My fault for dropping SWIFT_ENABLE_TARGET_FREEBSD without verifying that it didn't break lib/Driver/CMakeLists.txt target define handling.

How did you want to handle this for Linux? Should we just enable both targets by default? Or should I bring back SWIFT_ENABLE_TARGET_FREEBSD and make the tests conditional on the targets being enabled?

**Comment by gribozavr** [comment]
@landonf I'd prefer that we drop the linux option, too.

**Comment by gribozavr** [comment]
@landonf In fact, I'll do it now, give me an hour.

**Comment by jrose-apple** [comment]
In the spirit of [incremental development](https://swift.org/contributing/#incremental-development), can we do these as separate pull requests? Drop SWIFT_ENABLE_TARGET_LINUX, rename toolchain and fix-up the Driver library in preparation for the main change, then actually add the support. Or whatever units make sense to you. (Feel free to rewrite history ahead of the merges.)

Aside: Sorry for the trouble, Landon. I'm/We're very happy to have you working on this even as I'm pushing for this extra process. Having small, logically distinct commits is especially important if we ever need to bisect a failure—it's better to be able to isolate a specific change rather than potentially having to revert the whole merge commit. Also, once a change is in master, others are held accountable for not breaking it. :-)

**Comment by gribozavr** [comment]
I'm dropping SWIFT_ENABLE_TARGET_LINUX in a separate change.

**Comment by gribozavr** [comment]
Removed `SWIFT_ENABLE_TARGET_LINUX` in https://github.com/apple/swift/commit/a3c92e1b0da880f554f9353fecdceb10bebb5bef.  Unfortunately this created merge conflicts.  Sorry for the churn @landonf .

**Comment by landonf** [comment]
@gribozavr Thanks! No problem; I've merged in the changes (and adopted @jrose-apple's suggested "GenericUnix").

**Comment by gribozavr** [comment]
Running tests again.

**Comment by gribozavr** [comment]
@landonf The `-target` option accepts a triple.  Please apply a patch like this to your branch (use a more correct triple if required):

```patch
diff --git a/test/BuildConfigurations/x64FreeBSDTarget.swift b/test/BuildConfigurations/x64FreeBSDTarget.swift
index 619d4f5..284060f 100644
--- a/test/BuildConfigurations/x64FreeBSDTarget.swift
+++ b/test/BuildConfigurations/x64FreeBSDTarget.swift
@@ -1,5 +1,5 @@
-// RUN: %swift -parse %s -verify -D FOO -D BAR -target x86_64-freebsd10 -disable-objc-interop -D FOO -parse-stdlib
-// RUN: %swift-ide-test -test-input-complete -source-filename=%s -target x86_64-freebsd10
+// RUN: %swift -parse %s -verify -D FOO -D BAR -target x86_64-unknown-freebsd10 -disable-objc-interop -D FOO -parse-stdlib
+// RUN: %swift-ide-test -test-input-complete -source-filename=%s -target x86_64-unknown-freebsd10

 #if arch(x86_64) && os(FreeBSD) && _runtime(_Native)
 class C {}
```

**Comment by landonf** [comment]
Thanks, will do; I'm going to bring up a Linux machine so I can properly test locally.

**Comment by landonf** [comment]
@gribozavr Great; patchset now matches your change.

File: lib/Driver/Driver.cpp [hunk 1]
```diff
@@ -2036,7 +2036,10 @@ const ToolChain *Driver::getToolChain(const ArgList &Args) const {
       TC = new toolchains::Darwin(*this, Target);
       break;
     case llvm::Triple::Linux:
-      TC = new toolchains::Linux(*this, Target);
+      TC = new toolchains::GenericUnix(*this, Target);
+      break;
+    case llvm::Triple::FreeBSD:
```
Comment (archshift) [inline lib/Driver/Driver.cpp]: Why not just let this fall through instead of duplicating code?

Comment (gribozavr) [inline lib/Driver/Driver.cpp]: Fixed via https://github.com/apple/swift/commit/109d8e64027e2549a25f43ebe295879c1265863f, thanks!

**Comment by krzyzanowskim** [comment]
Does this change means that most `#if os(Linux)` should be now `#if os(Linux) || os(FreeBSD)` and all `else` conditions are undefined now?

`#if os(OSX) || os(iOS) || os(watchOS) || os(tvOS)
import Darwin
#elseif os(Linux)
import Glibc
#endif
`

ref: https://github.com/apple/swift/blob/8d9ef80304d7b36e13619ea50e6e76f3ec9221ba/stdlib/private/SwiftPrivatePthreadExtras/PthreadBarriers.swift#L13

**Comment by gribozavr** [comment]
Yes, all these places need to be audited (we tried to write them in such a way that a new platform would cause build breakage rather than silently choosing *some* behavior), but not all of them would become `os(Linux) || os(FreeBSD)`.  For example, FreeBSD does not use Glibc, it uses BSD libc, so the module likely wouldn't be called `Glibc`.

**Pull request merged**